### PR TITLE
DTS Dash Control GUI - Initial PR

### PR DESCRIPTION
**Pull request by colton-smith**
### DTS Dash Control GUI

This PR implements a dash GUI layout used for configuring, starting, and stopping tests on Paradigm's Dynamic Test Stand at Memorial University. A short demo of the working GUI can be seen in the#dts channel of the paradigm slack.

#### Whats in the box?
So far, only the necessary classes and callback functions for configuring and exporting test profiles have been implemented. In the future - functionality for starting a test (and less critically - importing test profiles) will be implemented. Also, the data visualization will be implemented using either dash or Grafana in the future.

The readme at Pidaq/UI provides insight into the project structure used for the dash application. Essentially, the dts dash layout and callbacks are packaged together in a package called `dts_dash`. They exist as modules in the subpackages `dts_dash.callbacks` and `dts_dash.layout`. This structure was chosen for scalability - if the UI grew to use many callbacks and or many different layout elements - everything can be segregated in an organized manner.

The  `dts_dash` package also includes an `app.py` file that simply declares the dash application. This app must get imported by the callback code - and so it must be declared here (before the callbacks are imported by the entrypoint/main - `dts-dash.py`) to avoid circular imports. This would not be possible if the app was declared in the insertion point of the program.

Utility classes are included such as `DtsCommand`, `DtsTestProfile`, and `DtsTestType(Enum)`. They are used when creating, editing, or exporting a DTS test profile via the GUI buttons. They are defined in the module `dts.py` in the package `Lib/utils`.

#### Summary
* `Lib/dts.py` - Module containing utility classes, functions, enums for the dts dash UI
* `dts_dash` - Package containing layout and callback modules for the dash UI
    * `callbacks/control_callbacks.py` - callbacks for the control GUI
    * `layout/control.py` - layout for the control GUI
    * `layout/manual.py` - barebones layout for the manual tab of the gui
    * `app.py` - creates dash app 
* `dts-dash.py` - Entrypoint for the dts dash application, imports the dash app from `dts_dash.dts`
* `README.md` - Describes how to launch the application and includes a directory structure diagram for the dash application/

**Note:** I have added the bootstrap css framework files to the repo, they can be ignored. The only css file that is user defined is `custom.css`.

### CODE REVIEW

File: Pidaq/Lib/utils/dts.py
Comment (colton-smith) [inline Pidaq/Lib/utils/dts.py]: Remove extra newline at eof

File: Pidaq/UI/DTS/dts-dash.py
Comment (colton-smith) [inline Pidaq/UI/DTS/dts-dash.py]: Should go after `from dts_dash.app import app` for correct alphabetical sorting

File: Pidaq/UI/DTS/dts_dash/callbacks/control_callbacks.py
Comment (colton-smith) [inline Pidaq/UI/DTS/dts_dash/callbacks/control_callbacks.py]: A lot of lines here exceed 80 chars - I was thinking fixing it may sacrifice some readability but I'm considering it

File: Pidaq/UI/DTS/dts_dash/callbacks/control_callbacks.py
Comment (colton-smith) [inline Pidaq/UI/DTS/dts_dash/callbacks/control_callbacks.py]: Add eof newline

File: Pidaq/UI/DTS/dts_dash/layout/control.py
Comment (colton-smith) [inline Pidaq/UI/DTS/dts_dash/layout/control.py]: - Gotta add a newline at EOF
- Some of the gross inline styles are in here because it wasnt working when I used classes to update the same css attributes, it was so weird. The most prominent example is the rpm/torque button group positioning.
- For line spacing in the bootstrap layout, I went with 2 spaces between rows and 1 between columns. I can go back to one line between each row though - just thought it might allow you to easier distinguish between rows and result in better readability

File: Pidaq/UI/DTS/dts_dash/layout/manual.py
Comment (colton-smith) [inline Pidaq/UI/DTS/dts_dash/layout/manual.py]: Add newline in eof

File: Pidaq/UI/README.md
Comment (colton-smith) [inline Pidaq/UI/README.md]: Should probably sort alphabetically... Not sure what the standard is for directory diagrams

File: Pidaq/Lib/utils/dts.py
Comment (Burke-Daniel) [inline Pidaq/Lib/utils/dts.py]: and the ...
You sound like me with my docstrings haha

File: Pidaq/Lib/utils/dts.py
Comment (Burke-Daniel) [inline Pidaq/Lib/utils/dts.py]: Love the code hinting

File: Pidaq/Lib/utils/dts.py
Comment (Burke-Daniel) [inline Pidaq/Lib/utils/dts.py]: Not that it matters very much, but I've been using snake_case instead of camelCase for method names, does PEP8 have an opinion on this?

File: Pidaq/Lib/utils/dts.py
Comment (Burke-Daniel) [inline Pidaq/Lib/utils/dts.py]: Should there be a space after the : in the type hint? Ex `testType: DTS...`

File: Pidaq/UI/DTS/dts-dash.py
Comment (Burke-Daniel) [inline Pidaq/UI/DTS/dts-dash.py]: EOF newline

File: requirements.txt
Comment (Burke-Daniel) [inline requirements.txt:15]: Thanks for cleaning this up haha

File: Pidaq/UI/DTS/dts_dash/layout/control.py
Comment (Burke-Daniel) [inline Pidaq/UI/DTS/dts_dash/layout/control.py]: Start the param list on the first line like the others for consistency

File: Pidaq/UI/README.md
Comment (Burke-Daniel) [reply Pidaq/UI/README.md]: Yeah not a bad idea, not sure what the standard is either but nbd either way

File: Pidaq/UI/DTS/dts-dash.py
Comment (Burke-Daniel) [inline Pidaq/UI/DTS/dts-dash.py]: The import sensors_callbacks is throwing an error presumably because that module hasn't been created yet, should this be left out of the imports for now? It ran fine once I removed it

File: Pidaq/UI/DTS/dts_dash/callbacks/control_callbacks.py
Comment (Burke-Daniel) [reply Pidaq/UI/DTS/dts_dash/callbacks/control_callbacks.py]: Up to you, not a huge deal though

File: Pidaq/Lib/utils/dts.py
Comment (Burke-Daniel) [reply Pidaq/Lib/utils/dts.py]: Or is there not a space because of the default value

**Review by Burke-Daniel — CHANGES_REQUESTED** [review]
This looks great! Ran it on my computer and ran into an issue with importing that sensor callback module that is yet to be defined like we spoke about, but after that I was able to run with no issues. Logging is great, As for the actual UI code, I don't really know enough about dash to have an opinion lol, but nothing stood out to me in that regard with my limited knowledge. @afhassan I'm thinking your review will be really useful on this, as you are the only other member at the moment who has been working with dash. This is hype though as I can see us being able to get a full simulation demo of running a test once we get the simulators/interfaces sorted out!

File: Pidaq/Lib/utils/dts.py
Comment (colton-smith) [reply Pidaq/Lib/utils/dts.py]: `RIP` - resolved

File: Pidaq/Lib/utils/dts.py
Comment (colton-smith) [reply Pidaq/Lib/utils/dts.py]: Yeah snake_case is the correct convention. Not sure why I went with lower first camelCase that breezy saturday morning. Fixing this.

Also, i'm gonna use that convention for variable names.

File: Pidaq/Lib/utils/dts.py
Comment (colton-smith) [reply Pidaq/Lib/utils/dts.py]: There should be a space, I think i did this because there is typically no space for keyword args between the operator,=, and operand, like:
```
def func(a, b=1, c=2, c=3);

# Then the call also has no spaces on either side of =:
func(1, b=10, c=20)
```

However yeah i'll make it
```
testType: DtsTestType=DtsTestType.RPM
```

File: Pidaq/UI/DTS/dts_dash/layout/control.py
Comment (colton-smith) [reply Pidaq/UI/DTS/dts_dash/layout/control.py]: Very good eye!

File: Pidaq/UI/DTS/dts-dash.py
Comment (colton-smith) [reply Pidaq/UI/DTS/dts-dash.py]: Yeah I meant to remove it / replace it with manual_callbacks.

For now, i'm just removing it. I can implement a manual callbacks file later when we actually need a manual.

**Review by Burke-Daniel — APPROVED** [review]
Looks GTM to me, would be good to have another set of eyes on this as well before merging but nbd either way.

**Pull request merged**
